File: quickrtc-flutter-client/lib/services/device_service.dart
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:logger/logger.dart';

/// Service for managing MediaSoup device
class DeviceService {
  static final DeviceService _instance = DeviceService._internal();
  factory DeviceService() => _instance;
  DeviceService._internal();

  final Logger _logger = Logger();

  RTCPeerConnection? _device;
  Map<String, dynamic>? _rtpCapabilities;
  bool _isLoaded = false;

  /// Load device with router RTP capabilities
  Future<RTCPeerConnection> loadDevice(
      Map<String, dynamic> routerRtpCapabilities) async {
    try {
      _logger.d('Loading device with router RTP capabilities');

      // In flutter_webrtc, we use RTCPeerConnection instead of mediasoup Device
      // The device initialization is handled differently
      final Map<String, dynamic> configuration = {
        'iceServers': [],
        'sdpSemantics': 'unified-plan',
      };

      _device = await createPeerConnection(configuration);
      _rtpCapabilities = routerRtpCapabilities;
      _isLoaded = true;

      _logger.i('Device loaded successfully');
      return _device!;
    } catch (e) {
      _logger.e('Failed to load device: $e');
      rethrow;
    }
  }

  /// Get the device instance
  RTCPeerConnection? getDevice() {
    return _device;
  }

  /// Get RTP capabilities
  Map<String, dynamic>? getRtpCapabilities() {
    return _rtpCapabilities;
  }

  /// Check if device is loaded
  bool isLoaded() {
    return _isLoaded;
  }

  /// Reset device
  Future<void> reset() async {
    _logger.d('Resetting device');

    if (_device != null) {
      await _device!.close();
      _device = null;
    }

    _rtpCapabilities = null;
    _isLoaded = false;

    _logger.i('Device reset completed');
  }
}

File: quickrtc-react-client/src/api/deviceService.ts
import { Device } from "mediasoup-client";
import type { RtpCapabilities } from "mediasoup-client/lib/types";

/**
 * Service for managing MediaSoup Device
 * Handles device initialization and capability management
 */
export class DeviceService {
  private device: Device | null = null;

  /**
   * Initialize device with router RTP capabilities
   */
  public async loadDevice(
    routerRtpCapabilities: RtpCapabilities
  ): Promise<Device> {
    if (this.device) {
      return this.device;
    }

    this.device = new Device();
    await this.device.load({ routerRtpCapabilities });

    return this.device;
  }

  /**
   * Get current device instance
   */
  public getDevice(): Device | null {
    return this.device;
  }

  /**
   * Get device RTP capabilities
   */
  public getRtpCapabilities(): RtpCapabilities | undefined {
    return this.device?.rtpCapabilities;
  }

  /**
   * Check if device is loaded
   */
  public isLoaded(): boolean {
    return this.device !== null && this.device.loaded;
  }

  /**
   * Reset device (for cleanup)
   */
  public reset(): void {
    this.device = null;
  }
}

// Export singleton instance
export const deviceService = new DeviceService();
